src/Grpc.NetCore.HttpClient/Internal/HttpContentClientStreamWriter.cs

```diff
@@ -0,0 +1,127 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+
+namespace Grpc.NetCore.HttpClient.Internal
+{
+    internal class HttpContentClientStreamWriter<TRequest, TResponse> : IClientStreamWriter<TRequest>
+    {
+        private readonly GrpcCall<TRequest, TResponse> _call;
+        private readonly Task<Stream> _writeStreamTask;
+        private readonly TaskCompletionSource<bool> _completeTcs;
+        private readonly object _writeLock;
+        private Task _writeTask;
+
+        public HttpContentClientStreamWriter(GrpcCall<TRequest, TResponse> call, Task<Stream> writeStreamTask, TaskCompletionSource<bool> completeTcs)
+        {
+            _call = call;
+            _writeStreamTask = writeStreamTask;
+            _completeTcs = completeTcs;
+            _writeLock = new object();
+            WriteOptions = _call.Options.WriteOptions;
+        }
+
+        public WriteOptions WriteOptions { get; set; }
+
+        public Task CompleteAsync()
+        {
+            lock (_writeLock)
+            {
+                // Pending writes need to be awaited first
+                if (IsWriteInProgress)
+                {
+                    return Task.FromException(new InvalidOperationException("Cannot complete client stream writer because the previous write is in progress."));
+                }
+
+                // Notify that the client stream is complete
+                _completeTcs.TrySetResult(true);
```

Comment:
This is dispatching right?

Reply:
Pardon?

It is used here: https://github.com/grpc/grpc-dotnet/blob/9812a63ea2aca64b4059594465c35690ddf379db/src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs#L302-L323

This writer writes to a stream, and the stream comes from the HttpContent on the request. The HttpContent is considered complete when its SerializeToStream method finishes. This TCS holds it open.